### PR TITLE
feat: K8S IaC - GitOps with ArgoCD

### PR DESCRIPTION
**Pull request by tty47**
## Description 📜

Hello team!

First of all, let me apologize for the size of this PR 🙏 

TLDR:
*This PR contains a lot of changes in the way that we create and provision the Kubernetes clusters for TestGround.*

## Original approach: ⛳️
We started developing the IaC with Terraform (EKS) + [ArgoCD](https://argo-cd.readthedocs.io/en/stable/), but after some tests, we realized that EKS has some limitations when we need to run big clusters (>7k tests pods), this is because the CP ( K8S ControlPlane) is self-managed by AWS, and the amount of resources is managed by them.

## Then, we pivoted to: 🎯
After a while, we decided to move to [Kops](https://kops.sigs.k8s.io/), and following the same approach,  IaC + GitOps, we now have the cluster creation with [Kops](https://kops.sigs.k8s.io/) + [ArgoCD](https://argo-cd.readthedocs.io/en/stable/).

### Some benefits of using this approach:
- We manage the CP and it has autoscaling, so, no matter how many pods we want to spin up, the cluster will be able to support them.
- GitOps, everything as code, so the configuration is in the repository.
  - We can reuse the provisioning code, as we are using [ArgoCD](https://argo-cd.readthedocs.io/en/stable/), all the resources + tools are in the repo. 
- Portable, we can move to another cloud provider and reuse the provisioning code.
- EFS created with Terraform
- StorageClasses created by Terraform

ℹ️  *Also, mention that both clusters are functional, you can use either of them (EKS/Kops).*

Thanks in advance!

--- 

### What's included in this PR 🧐

- AMI updates:
  - Image reference updated
  - Script which installs some tools in the nodes fixed to be able to use EFS (client NFS) 
- Documentation
- Scripts to install & delete the cluster updated
- Remove files that are not needed anymore
- Upgrade to `v1.23.4` -> latest K8S version we can use while running Docker as CRI.
- Helm Charts added
- TestsGround tested

ℹ️ *Once we merge this PR, we will have to update all the references to this branch.*

---

Closes: https://github.com/celestiaorg/devops/issues/123

--- 

Best 🚀 

Jose Ramon Mañes

### CODE REVIEW

File: manifests/testground/daemon/tg-daemon-deployment.yaml [hunk 1]
```diff
@@ -48,16 +35,19 @@ spec:
...
+            memory: 700Mi
             cpu: 500m
           limits:
-            memory: 512Mi
+            memory: 1000Mi
+            cpu: 700m
       - name: testground-daemon
-        image: iptestground/testground:edge
+        #image: public.ecr.aws/n6b0k8i7/testground-daemon:80c5aca36114de067c33c8718cca95ef16db4c06
+        image: ghcr.io/celestiaorg/testground:39661ea
```
Comment (Bidon15) [inline manifests/testground/daemon/tg-daemon-deployment.yaml]: we will need to bump to latest version in order to make all changes work

File: terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      userarn  = "arn:aws:iam::506657148836:user/alexk@celestia.org"
+      username = "alexk@celestia.org"
+      groups   = ["system:masters"]
+    },
+    {
+      userarn  = "arn:aws:iam::506657148836:user/jose@celestia.org"
+      username = "jose@celestia.org"
+      groups   = ["system:masters"]
+    }
+  ]
```
Comment (Bidon15) [inline terraform/main.tf]: this is not the best option to include here in the PR

File: terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,415 @@
+//################################################################################
+//# Locals
+//################################################################################
+locals {
+  region       = "eu-west-1"
+  project_name = "devops-2"
```
Comment (Bidon15) [inline terraform/main.tf]: we can include a readme why eu-west-1 is best 

TL;DR - it has the most resources available for tg

File: terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+      //custom_ami_id      = "ami-083694f0a1d262109"
+      //create_launch_template = true              # false will use the default launch template
+      //launch_template_os = "amazonlinux2eks" # amazonlinux2eks/bottlerocket # Used to identify the launch template
+
+      // TODO: FIXME PLEASE!
+      // we are hardcoding the first element of the list in order to use ONLY
+      // the first az in the nodes
+      //subnet_ids      = split(",", module.vpc.public_subnets[0])
+      //subnets_id = element(module.vpc.public_subnets, 0)
+      //subnets_id = "subnet-0315baa5c81ff7d10"
```
Comment (Bidon15) [inline terraform/main.tf]: can be removed?

File: terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+      //LIMITS'
+      //EOT
+
+      //      iam:
+      // attachPolicyARNs:
+      // - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess
+      // - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
+      // - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
+      // - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
+      // - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
```
Comment (Bidon15) [inline terraform/main.tf]: can this be removed?

File: terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+      //net.ipv4.neigh.default.gc_thresh3 = 32768
+      //SYSCTL'
+      //sudo sysctl -p /etc/sysctl.d/999-testground.conf"
+      //sudo bash -c 'cat <<LIMITS > /etc/security/limits.d/999-limits.conf
+      //* soft nproc 131072
+      //* hard nproc 262144
+      //* soft nofile 131072
+      //* hard nofile 262144
+      //LIMITS'
+      //EOT
```
Comment (Bidon15) [inline terraform/main.tf]: same

File: terraform/main.tf [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+      path               = "chart"
+      repo_url           = "https://github.com/aws-samples/eks-blueprints-add-ons.git"
+      add_on_application = true # Indicates the root add-on application.
+    }
+
+    // https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/modules/kubernetes-addons/argocd/main.tf#L94-L114
+    kustomize-apps = {
+      path            = "argocd-root"
+      repo_url        = "https://github.com/celestiaorg/testground-infra.git"
+      target_revision = "jose/hackground-k8s-tf" // Remove this one, should use main
```
Comment (Bidon15) [inline terraform/main.tf:311]: we need to prepare this to use testground/testground after this PR get's merged in out celestiaorg/testground and we do a proper PR to upstream

**Review by Bidon15 — APPROVED** [review]
Fantastic job!

**Review by Bidon15 — COMMENTED** [review]
Fantastic job!

**Comment by Bidon15** [comment]
> Fantastic job!

Discard my approve :D 
I mispressed the button 😓

File: terraform/main.tf (hunk 2 above)
Comment (tty47) [reply terraform/main.tf]: good catch! I missed it, thank you!

File: terraform/main.tf (hunk 5 above)
Comment (tty47) [reply terraform/main.tf]: I just uncommented it, I believe we do need it, the cluster works with and without it, but for preventing any issues, I'll keep it there

File: manifests/testground/daemon/tg-daemon-deployment.yaml (hunk 1 above)
Comment (tty47) [reply manifests/testground/daemon/tg-daemon-deployment.yaml]: updated to the latest available commit: `900c23e`


https://github.com/celestiaorg/testground/pkgs/container/testground/79647287?tag=900c23e

File: terraform/main.tf (hunk 7 above)
Comment (tty47) [reply terraform/main.tf:311]: exactly, I had the same idea, after we merge this PR, we'll be able to test it using the celestia repo, if everything works (it should), then, I'll prepare a new PR pointing to the final repo: [testground/infra](https://github.com/testground/infra)

**Review by Bidon15 — APPROVED** [review]
Let's merge this and do a PR in testground/infra
